File: backend/app/services/news_crawler.py
import httpx
from datetime import datetime
from typing import Optional
from app.models import NewsArticle
from app.utils.id_generator import generate_news_id
from app.utils.date_parser import parse_published_date


class NewsCrawler:
    def __init__(self, api_key: str):
        self.api_key = api_key
        self.base_url = "https://serpapi.com/search.json"

    async def search_news(
        self,
        query: str,
        language: str = "ko",
        country: str = "kr",
        num: int = 100
    ) -> list[NewsArticle]:
        """
        Search news using SerpAPI Google News engine.
        Supports fetching more than 100 results by making multiple paginated requests.
        num: Total number of results to return (will make multiple API calls if > 100)
        """
        all_articles = []
        per_page = 100  # SerpAPI limit per request
        pages_needed = (num + per_page - 1) // per_page  # Ceiling division

        async with httpx.AsyncClient(timeout=30.0) as client:
            for page in range(pages_needed):
                start = page * per_page
                current_num = min(per_page, num - start)

                if current_num <= 0:
                    break

                params = {
                    "engine": "google_news",
                    "q": query,
                    "hl": language,
                    "gl": country,
                    "num": current_num,
                    "start": start,
                    "api_key": self.api_key
                }

                try:
                    response = await client.get(self.base_url, params=params)
                    response.raise_for_status()
                    data = response.json()
                except httpx.HTTPError as e:
                    # If pagination fails, return what we have so far
                    if page > 0:
                        break
                    raise Exception(f"SerpAPI request failed: {str(e)}")

                # Extract news_results array
                news_results = data.get("news_results", [])

                # If no more results, stop fetching
                if not news_results:
                    break

                for item in news_results:
                    article = self._parse_news_item(item)
                    if article:
                        all_articles.append(article)

        # Remove duplicates based on ID
        seen_ids = set()
        unique_articles = []
        for article in all_articles:
            if article.id not in seen_ids:
                seen_ids.add(article.id)
                unique_articles.append(article)

        # Sort by publishedAt (None values go to the back)
        from datetime import timezone
        def get_sort_key(article):
            if article.publishedAt:
                if article.publishedAt.tzinfo is None:
                    return article.publishedAt.replace(tzinfo=timezone.utc)
                return article.publishedAt
            return datetime(1970, 1, 1, tzinfo=timezone.utc)

        unique_articles.sort(key=get_sort_key, reverse=True)

        return unique_articles

    def _parse_news_item(self, item: dict) -> Optional[NewsArticle]:
        """Parse a single news item from SerpAPI response."""
        try:
            title = item.get("title", "")
            url = item.get("link", "")

            if not title or not url:
                return None

            # Generate ID
            article_id = generate_news_id(url, title)

            # Parse date - try multiple possible field names
            date_str = item.get("date") or item.get("time") or item.get("published") or ""

            # Extract source name
            source_info = item.get("source", {})
            if isinstance(source_info, dict):
                source_name = source_info.get("name", "google")
            else:
                source_name = str(source_info) if source_info else "google"

            # Parse published date
            published_at = None
            if date_str:
                published_at = parse_published_date(date_str, source_name)

            return NewsArticle(
                id=article_id,
                title=title,
                url=url,
                source=source_name,
                publishedAt=published_at,
                snippet=item.get("snippet"),
                thumbnail=item.get("thumbnail")
            )
        except Exception:
            return None

File: backend/src/services/newsCrawler.js
const Parser = require('rss-parser');
const { generateNewsId } = require('../utils/idGenerator');
const { parsePublishedDate } = require('../utils/dateParser');

class NewsCrawler {
  constructor() {
    this.parser = new Parser({
      timeout: 15000,
      headers: {
        'User-Agent':
          'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
      },
    });
  }

  /**
   * Search news using Google News RSS feed.
   * Fetches from multiple time-scoped queries to maximize results.
   * @param {string} query - Search query
   * @param {string} hl - Language code (e.g. 'ko')
   * @param {string} gl - Country code (e.g. 'kr')
   * @param {number} num - Maximum number of articles
   * @returns {Promise<Array>} - Array of article objects
   */
  async searchNews(query, hl = 'ko', gl = 'kr', num = 500) {
    const encodedQuery = encodeURIComponent(query);
    const ceid = `${gl.toUpperCase()}:${hl}`;

    // Fetch from multiple Google News RSS URLs to maximize results
    const base = `hl=${hl}&gl=${gl.toUpperCase()}&ceid=${ceid}`;
    const rssUrls = [
      // Default (recent)
      `https://news.google.com/rss/search?q=${encodedQuery}&${base}`,
      // Time ranges
      `https://news.google.com/rss/search?q=${encodedQuery}+when:1d&${base}`,
      `https://news.google.com/rss/search?q=${encodedQuery}+when:3d&${base}`,
      `https://news.google.com/rss/search?q=${encodedQuery}+when:7d&${base}`,
      `https://news.google.com/rss/search?q=${encodedQuery}+when:30d&${base}`,
      // If multi-word, also search with quotes for exact match
      ...(query.includes(' ') ? [
        `https://news.google.com/rss/search?q=%22${encodedQuery}%22&${base}`,
      ] : []),
      // Also try English locale for international coverage
      ...(hl !== 'en' ? [
        `https://news.google.com/rss/search?q=${encodedQuery}&hl=en&gl=US&ceid=US:en`,
      ] : []),
    ];

    const allArticles = [];

    const feedPromises = rssUrls.map(async (rssUrl) => {
      try {
        const feed = await this.parser.parseURL(rssUrl);
        const items = feed.items || [];
        const articles = [];
        for (const item of items) {
          const article = this._parseRssItem(item);
          if (article) articles.push(article);
        }
        return articles;
      } catch {
        return [];
      }
    });

    const results = await Promise.allSettled(feedPromises);
    for (const result of results) {
      if (result.status === 'fulfilled' && Array.isArray(result.value)) {
        allArticles.push(...result.value);
      }
    }

    // Deduplicate
    const seenIds = new Set();
    const unique = [];
    for (const article of allArticles) {
      if (!seenIds.has(article.id)) {
        seenIds.add(article.id);
        unique.push(article);
      }
    }

    // Sort newest first
    unique.sort((a, b) => {
      const dateA = a.publishedAt ? new Date(a.publishedAt).getTime() : 0;
      const dateB = b.publishedAt ? new Date(b.publishedAt).getTime() : 0;
      return dateB - dateA;
    });

    return unique.slice(0, num);
  }

  _parseRssItem(item) {
    try {
      const title = (item.title || '').trim();
      const rawLink = item.link || '';
      if (!title || !rawLink) return null;

      // Extract real URL from Google redirect
      const url = this._extractRealUrl(rawLink);

      // Source: Google RSS title format is "Article Title - Source Name"
      let sourceName = 'google';
      if (item.source) {
        sourceName = item.source;
      } else {
        const dashIdx = title.lastIndexOf(' - ');
        if (dashIdx > 0) {
          sourceName = title.substring(dashIdx + 3).trim();
        }
      }

      // Clean title (remove " - Source Name" suffix if present)
      let cleanTitle = title;
      const dashIdx = title.lastIndexOf(' - ');
      if (dashIdx > 0) {
        cleanTitle = title.substring(0, dashIdx).trim();
      }

      const publishedAt = item.pubDate
        ? parsePublishedDate(item.pubDate, sourceName)
        : null;

      const articleId = generateNewsId(url, cleanTitle);

      return {
        id: articleId,
        title: cleanTitle,
        url,
        source: sourceName,
        publishedAt: publishedAt ? publishedAt.toISOString() : null,
        snippet: this._cleanSnippet(item.contentSnippet || item.content || null, sourceName),
        thumbnail: null,
      };
    } catch {
      return null;
    }
  }

  _cleanSnippet(snippet, source) {
    if (!snippet) return null;
    let text = snippet.replace(/\u00a0/g, ' ').trim();
    if (source) {
      const re = new RegExp(`\\s*${source.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\s*$`);
      text = text.replace(re, '').trim();
    }
    return text || null;
  }

  /**
   * Extract the real URL from a Google News redirect URL.
   * Google News RSS links look like:
   *   https://news.google.com/rss/articles/CBMi...
   * or contain a url= query parameter.
   */
  _extractRealUrl(googleUrl) {
    try {
      const parsed = new URL(googleUrl);

      // Check for url query parameter
      const urlParam = parsed.searchParams.get('url');
      if (urlParam) return urlParam;

      // If it's a Google News redirect, return as-is (will redirect on click)
      return googleUrl;
    } catch {
      return googleUrl;
    }
  }
}

module.exports = { NewsCrawler };
